### PR TITLE
End-to-end "in memory" tests-simple-study : simplify / clarify

### PR DESCRIPTION
**Pull request by guilpier-code**

### CODE REVIEW

File: src/tests/end-to-end/utils/utils.h [hunk 1]
```diff
@@ -8,11 +8,49 @@
 
 using namespace Antares::Solver;
 using namespace Antares::Solver::Simulation;
+using namespace Antares::Data::ScenarioBuilder;
 
 
 void initializeStudy(Study::Ptr study);
-void setNumberMCyears(Study::Ptr study, unsigned int nbYears);
 void configureLinkCapacities(AreaLink* link);
+void addLoadToArea(Area* area, double loadInArea);
```
Comment (sylvlecl) [inline src/tests/end-to-end/utils/utils.h]: I think we should take the habit of using references where argument must not be null, instead of pointer.

File: src/tests/end-to-end/utils/utils.h [hunk 2]
```diff
@@ -8,11 +8,49 @@
...
+
+
+template<class MatixType>
```
Comment (sylvlecl) [inline src/tests/end-to-end/utils/utils.h]: typo in Mat**r**ixType

File: src/tests/end-to-end/binding_constraints/test_binding_constraints.cpp [hunk 3]
```diff
@@ -63,18 +30,25 @@ StudyForBCTest::StudyForBCTest()
     Area* area1 = addAreaToStudy("Area 1");
     Area* area2 = addAreaToStudy("Area 2");
 
-    double loadInAreaOne = 0.;
-    addLoadToArea(area1, loadInAreaOne);
+    auto loadTSconfig = std::make_shared<TimeSeriesConfig<Matrix<double, int32_t>>>(area1->load.series->timeSeries);
```
Comment (sylvlecl) [inline src/tests/end-to-end/binding_constraints/test_binding_constraints.cpp]: There is no reason for doing heap allocation and using a shared_ptr here, let's use a plain value on the stack.

File: src/tests/end-to-end/utils/utils.h [hunk 4]
```diff
@@ -8,11 +8,49 @@
...
+class TimeSeriesConfig
+{
+public:
+    TimeSeriesConfig(MatixType& matrix) : ts_(&matrix) {}
+    void setTargetTSmatrix(MatixType& matrix) { ts_ = &matrix; }
```
Comment (sylvlecl) [inline src/tests/end-to-end/utils/utils.h]: I would propose to remove that possibility: the matrix is already provided in the constructor, if we want to change the matrix we can just use a new configurer.

File: src/tests/end-to-end/utils/utils.h [hunk 5]
```diff
@@ -8,11 +8,49 @@
...
 
 
 void initializeStudy(Study::Ptr study);
-void setNumberMCyears(Study::Ptr study, unsigned int nbYears);
 void configureLinkCapacities(AreaLink* link);
+void addLoadToArea(Area* area, double loadInArea);
+
+
+template<class MatixType>
+class TimeSeriesConfig
```
Comment (sylvlecl) [inline src/tests/end-to-end/utils/utils.h]: Maybe "configur**er**" ? The object is not a configuration, it's something that is used to configure the matrix.

File: src/tests/end-to-end/utils/utils.h [hunk 6]
```diff
@@ -8,11 +8,49 @@
...
+{
+public:
+    TimeSeriesConfig(MatixType& matrix) : ts_(&matrix) {}
+    void setTargetTSmatrix(MatixType& matrix) { ts_ = &matrix; }
+    void setNumberColumns(unsigned int nbColumns) { ts_->resize(nbColumns, HOURS_PER_YEAR);}
```
Comment (sylvlecl) [inline src/tests/end-to-end/utils/utils.h]: I think "columns count" would be a better name.

File: src/tests/end-to-end/utils/utils.h [hunk 7]
```diff
@@ -8,11 +8,49 @@
...
+    void fillColumnWith(unsigned int column, double value) { ts_->fillColumn(column, value); }
+private:
+    MatixType* ts_;
+};
+
+class ThermalClusterConfig
+{
+public:
+    ThermalClusterConfig(std::shared_ptr<ThermalCluster> cluster) : cluster_(cluster)
```
Comment (sylvlecl) [inline src/tests/end-to-end/utils/utils.h]: No need for shared_ptr here since the configurer will not own the cluster.

File: src/tests/end-to-end/utils/utils.h [hunk 8]
```diff
@@ -8,11 +8,49 @@
...
+        cluster_->marginalCost = cost;
+        cluster_->marketBidCost = cost; // Must define market bid cost otherwise all production is used
+        cluster_->setProductionCost();
+    }
+    void setAvailablePowerNumberOfTS(unsigned int nbColumns) { tsAvailablePowerConfig_->setNumberColumns(nbColumns); };
+    void setAvailablePower(unsigned int column, double value) { tsAvailablePowerConfig_->fillColumnWith(column, value); }
+
+private:
+    std::shared_ptr<ThermalCluster> cluster_;
+    std::unique_ptr<TimeSeriesConfig<Matrix<double>>> tsAvailablePowerConfig_;
```
Comment (sylvlecl) [inline src/tests/end-to-end/utils/utils.h]: Here also I think there is no point in doing heap allocation for this object, we can have a plain value.

File: src/tests/end-to-end/binding_constraints/test_binding_constraints.cpp [hunk 9]
```diff
@@ -117,136 +91,97 @@ BOOST_FIXTURE_TEST_SUITE(TESTS_BINDING_CONSTRAINTS_ON_A_LINK, StudyWithBConLink)
 
 BOOST_AUTO_TEST_CASE(Hourly_BC_restricts_link_direct_capacity_to_90)
```
Comment (sylvlecl) [inline src/tests/end-to-end/binding_constraints/test_binding_constraints.cpp:92]: Not part of this PR, but I think we should use test **case** fixtures, not test **suite** fixtures, to better isolate unit tests.
Study should be create from scratch for each unit test, and deleted at the end of the unit test.

File: src/tests/end-to-end/binding_constraints/test_binding_constraints.cpp (hunk 3 above)
Comment (sylvlecl) [reply src/tests/end-to-end/binding_constraints/test_binding_constraints.cpp]: For this one and other classes, we can use the classic builder pattern for chaining calls and not repeat the config each time:
```cpp
    TimeSeriesConfig(area1->load.series->timeSeries)
        .setNumberColumns(1)
        .fillColumnWith(0, 0);

    TimeSeriesConfig(area2->load.series->timeSeries)
        .setNumberColumns(1)
        .fillColumnWith(0, 100);
```

**Review by sylvlecl — COMMENTED** [review]
Some suggestions for improvement.

File: src/tests/end-to-end/utils/utils.h (hunk 1 above)
Comment (guilpier-code) [reply src/tests/end-to-end/utils/utils.h]: Unused function, was removed

File: src/tests/end-to-end/utils/utils.h (hunk 2 above)
Comment (guilpier-code) [reply src/tests/end-to-end/utils/utils.h]: Done (as soon as next push is made)

File: src/tests/end-to-end/utils/utils.h [hunk 10]
```diff
@@ -8,13 +8,64 @@
...
+
+template<class MatrixType>
+class TimeSeriesConfig
+{
+public:
+    TimeSeriesConfig(MatrixType& matrix) : ts_(&matrix) {}
+    TimeSeriesConfig& setNumberColumns(unsigned int nbColumns);
+    TimeSeriesConfig& fillColumnWith(unsigned int column, double value);
+private:
+    MatrixType* ts_;
```
Comment (flomnes) [inline src/tests/end-to-end/utils/utils.h]: Maybe it would be nice to have a common `IMatrix` interface, so we don't need to templatize everything.

Something like
```cpp
class IMatrix {
    virtual std::size_t width() const;
    virtual std::size_t height() const;
    virtual void resize(std::size_t width, std::size_t height);
};

template<class T, class U>
class Matrix : public IMatrix {
 // ...
};
```

File: src/tests/end-to-end/utils/utils.h (hunk 4 above)
Comment (guilpier-code) [reply src/tests/end-to-end/utils/utils.h]: Done

File: src/tests/end-to-end/utils/utils.h (hunk 7 above)
Comment (guilpier-code) [reply src/tests/end-to-end/utils/utils.h]: Done

File: src/tests/end-to-end/utils/utils.h (hunk 6 above)
Comment (guilpier-code) [reply src/tests/end-to-end/utils/utils.h]: Done

File: src/tests/end-to-end/utils/utils.h (hunk 5 above)
Comment (guilpier-code) [reply src/tests/end-to-end/utils/utils.h]: Done

File: src/tests/end-to-end/utils/utils.h (hunk 8 above)
Comment (guilpier-code) [reply src/tests/end-to-end/utils/utils.h]: Done. For both the cluster and the TS configurer.

File: src/tests/end-to-end/binding_constraints/test_binding_constraints.cpp (hunk 3 above)
Comment (guilpier-code) [reply src/tests/end-to-end/binding_constraints/test_binding_constraints.cpp]: Done, as much as I could.

File: src/tests/end-to-end/utils/utils.h (hunk 10 above)
Comment (guilpier-code) [reply src/tests/end-to-end/utils/utils.h]: If I understand you properly, we would inject the **Matrix** in the **TimeSeriesConfigurer** at construction, meaning that the **TimeSeriesConfigurer**'s constructor would expect an **IMatrix** as parameter,  right ?

Comment (flomnes) [reply src/tests/end-to-end/utils/utils.h]: You'd inject it's `IMatrix` part through [object slicing](https://en.wikipedia.org/wiki/Object_slicing)

**Review by sylvlecl — APPROVED** [review]
Nice, I'll just remove template arguments that can be deduced before merging in the other branch.

**Review by flomnes — APPROVED** [review]
